File: pathfinder-api/src/main/java/de/cubbossa/pathapi/storage/cache/VisualizerCache.java
package de.cubbossa.pathapi.storage.cache;

import de.cubbossa.pathapi.misc.NamespacedKey;
import de.cubbossa.pathapi.visualizer.PathVisualizer;
import de.cubbossa.pathapi.visualizer.VisualizerType;

import java.util.Collection;
import java.util.Optional;

public interface VisualizerCache extends StorageCache<PathVisualizer<?, ?>> {

  <VisualizerT extends PathVisualizer<?, ?>> Optional<VisualizerT> getVisualizer(NamespacedKey key);

  Optional<Collection<PathVisualizer<?, ?>>> getVisualizers();

  <VisualizerT extends PathVisualizer<?, ?>> Optional<Collection<VisualizerT>> getVisualizers(VisualizerType<VisualizerT> type);

  <VisualizerT extends PathVisualizer<?, ?>> void writeAll(VisualizerType<VisualizerT> type, Collection<VisualizerT> v);

  void writeAll(Collection<PathVisualizer<?, ?>> visualizers);
}

File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/storage/cache/VisualizerCache.kt
package de.cubbossa.pathfinder.storage.cache

import de.cubbossa.pathfinder.misc.NamespacedKey
import de.cubbossa.pathfinder.visualizer.PathVisualizer
import de.cubbossa.pathfinder.visualizer.VisualizerType
import java.util.*

interface VisualizerCache : StorageCache<PathVisualizer<*, *>> {
    fun <VisualizerT : PathVisualizer<*, *>> getVisualizer(key: NamespacedKey): VisualizerT?

    val visualizers: Optional<Collection<PathVisualizer<*, *>>>

    fun <VisualizerT : PathVisualizer<*, *>> getVisualizers(type: VisualizerType<VisualizerT>): Collection<VisualizerT>?

    fun <VisualizerT : PathVisualizer<*, *>> writeAll(
        type: VisualizerType<VisualizerT>,
        v: Collection<VisualizerT>
    )

    fun writeAll(visualizers: Collection<PathVisualizer<*, *>>)
}
